core/sessions/oidcauth/oidc.go

```diff
@@ -0,0 +1,675 @@
+	exi := &bridges.ExternalInitiator{}
+	err := oi.ds.GetContext(ctx, exi, `SELECT * FROM external_initiators WHERE access_key = $1`, eia.AccessKey)
+	return exi, err
+}
+
+// localLoginFallback tests the credentials provided against the 'local' authentication method
+// This covers the case of local CLI API calls requiring local login separate from the OIDC server
+func (oi *oidcAuthenticator) localLoginFallback(ctx context.Context, sr clsessions.SessionRequest) (clsessions.User, error) {
+	var user clsessions.User
+	err := oi.ds.GetContext(ctx, &user, SQLSelectUserbyEmail, sr.Email)
+	if err != nil {
+		return user, err
+	}
+	if !constantTimeEmailCompare(strings.ToLower(sr.Email), strings.ToLower(user.Email)) {
+		oi.auditLogger.Audit(audit.AuthLoginFailedEmail, map[string]interface{}{"email": sr.Email})
+		return user, errors.New("invalid email")
+	}
+
+	if !utils.CheckPasswordHash(sr.Password, user.HashedPassword) {
+		oi.auditLogger.Audit(audit.AuthLoginFailedPassword, map[string]interface{}{"email": sr.Email})
+		return user, errors.New("invalid password")
+	}
+
+	return user, nil
+}
+
+func (oi *oidcAuthenticator) IDClaimsToUserRole(idClaims []string, adminClaim string, editClaim string, runClaim string, readClaim string) (clsessions.UserRole, error) {
+	// If defined Admin group name is present in id claims, return UserRoleAdmin
+	for _, group := range idClaims {
+		if group == adminClaim {
+			return clsessions.UserRoleAdmin, nil
+		}
+	}
+	// Check edit role
+	for _, group := range idClaims {
+		if group == editClaim {
+			return clsessions.UserRoleEdit, nil
+		}
+	}
+	// Check run role
+	for _, group := range idClaims {
+		if group == runClaim {
+			return clsessions.UserRoleRun, nil
+		}
+	}
+	// Check view role
+	for _, group := range idClaims {
+		if group == readClaim {
+			return clsessions.UserRoleView, nil
+		}
+	}
+	// No role group found, error
+	return clsessions.UserRoleView, ErrUserNoOIDCGroups
+}
+
+// extractIDClaimValues extracts groups from the claims using the specified key
+func (oi *oidcAuthenticator) ExtractIDClaimValues(claims map[string]interface{}, key string) ([]string, error) {
+	claimValues, ok := claims[key]
+	if !ok {
+		return nil, fmt.Errorf("claim '%s' not found in ID token", key)
+	}
+
+	// Handle different types of claim values
+	switch v := claimValues.(type) {
+	case []interface{}:
+		val := make([]string, 0, len(v))
+		for _, item := range v {
+			str, ok := item.(string)
+			if !ok {
+				return nil, fmt.Errorf("invalid type for item in '%s': expected string, got %T", key, item)
+			}
+			val = append(val, str)
+		}
+		return val, nil
+	case []string:
+		return v, nil
+	case string:
+		return []string{v}, nil
+	default:
+		return nil, fmt.Errorf("claim '%s' is not a string or array: got %T", key, v)
+	}
+}
+
+const constantTimeEmailLength = 256
+
+func constantTimeEmailCompare(left, right string) bool {
+	length := mathutil.Max(constantTimeEmailLength, len(left), len(right))
```

Comment:
nit/ this is valid if you want to declare it in the narrowest scope:
```suggestion
func constantTimeEmailCompare(left, right string) bool {
	const constantTimeEmailLength = 256
	length := mathutil.Max(constantTimeEmailLength, len(left), len(right))```

Reply:
Nice, done!